### DISCUSSION TITLE
Laravel Echo?

### DISCUSSION THREAD

**Discussion by defunctl** [filamentphp/filament#1011] [Help]
I'm trying to implement Laravel Echo alongside of Filament, however I'm not finding the correct place to either inject the Echo JS or a way to override base.blade.php to inject it in the head (anywhere else and it seems LaravelEcho.js loads before the configuration and is unable to find the `window.Echo` global.

Any suggestions in getting this working?

Thanks

Edit: just trying to run this JS before all the rest of the JavaScript is loaded:

```JavaScript
import Echo from 'laravel-echo';

window.Pusher = require('pusher-js');

window.Echo = new Echo({
    broadcaster: 'pusher',
    key: process.env.MIX_PUSHER_APP_KEY,
    wsHost: '127.0.0.1',
    wsPort: 6001,
    forceTLS: false,
    encrypted: true,
    disableStats: true,
    enabledTransports: ['ws', 'wss'],
});
```

**Comment by danharrin** [comment]
Hmm, have you tried to `Filament::registerScripts(['echo' => 'path to echo here'])`

**Reply by defunctl** [reply]
@danharrin I'll give that a go and report back, I did register my main laravel app.js within a PluginServiceProvider as [per the docs](https://filamentadmin.com/docs/2.x/admin/plugins#scripts), but it was too late in the execution.

edit: looks like that uses `Filament::registerScripts()`  under the hood, so unfortunately that won't work.

**Reply by defunctl** [reply]
Basically need a `<script src="{{ mix('js/app.js') }}"></script>` before [this line](https://github.com/laravel-filament/filament/blob/2.x/packages/admin/resources/views/components/layouts/base.blade.php#L38) and it works.

**Reply by danharrin** [reply]
Hmm, do you need echo on just a few pages?

Maybe just add the script above the code where you require it, as that will always be above the footer of the page.

**Reply by defunctl** [reply]
Unfortunately it's needs to be on every admin page in order to check the pusher server for notifications to display.

https://laravel-livewire.com/docs/2.x/laravel-echo

**Comment by defunctl** [comment]
@danharrin I got around this for now by overriding the view path for the sidebar `start.blade.php` which is both empty and loads before `@livewireScripts` in the base view. 

It would actually be nice to get something like this in the base template?

Anyways, for anyone else with this problem, here is my service provider:

```php
<?php declare(strict_types=1);

namespace App\Providers;

use Illuminate\Support\Facades\View;
use Illuminate\Support\ServiceProvider;

class AppServiceProvider extends ServiceProvider {

    /**
     * Register any application services.
     *
     * @return void
     */
    public function register() {
        //
    }

    /**
     * Bootstrap any application services.
     *
     * @return void
     */
    public function boot() {

        /**
         * Override the Filament sidebar start.blade.php view to inject our "Laravel Echo" global
         * JavaScript before Livewire loads its LaravelEcho component.
         */
        View::creator( 'filament::components.layouts.app.sidebar.start', static function ( \Illuminate\View\View $view ) {
            $view->setPath( resource_path( 'views/filament/overrides/components/app/sidebar/start.blade.php' ) );
        } );
    }

}
```

And my start.blade.php:

```html
{{-- Add our application's JavaScript before other scripts are loaded by livewire --}}
<script src="{{ mix('js/app.js') }}"></script>
```

**Reply by hunterd** [reply]
Thanks for the workaround.
Actually, if you put a copy of start.blade.php in the `resources/views/vendor/filament/components/layouts/app/sidebar/start.blade.php` it automatically overrides the original file located in vendor.
Thanks to this method, I was able to add `@vite(['resources/js/bootstrap.js'])` in the start.blade.php and access to Laravel Echo

**Reply by danharrin** [reply]
Please dont override views... Laravel Echo is now supported out of the box, check out the config file

**Reply by hunterd** [reply]
Thank you for the information. I tried to make this work with the filament.php file configuration, but I always get "Laravel Echo cannot be found" message in the debug console.

bootstrap.php
```
'broadcasting' => [

        'echo' => [
            'broadcaster' => 'pusher',
            'key' => env('VITE_PUSHER_APP_KEY'),
            'wsHost' => preg_replace('/^(https?:\/\/)/', '', env('APP_URL')),
            'wsPort' => env('VITE_PUSHER_APP_PORT', 6001),
            'wssPort' => env('VITE_PUSHER_APP_PORT', 6001),
            'forceTLS' => true,
            'enabledTransports' => ['ws', 'wss'],
            'cluster' => env('VITE_PUSHER_APP_CLUSTER'),
        ],

    ],
```
Those informations are fulfilled like it was in "bootstrap.js" but since I'm working with Vite.js and that I have to work in HTTPS version of my site I have to pass this data to make it work with my Laravel Websockets instance :
 
vite.config.js 
 ```
server: {
    host: process.env.APP_URL.replace(/^(https?:\/\/)/, ''),
    https: {
        key: fs.readFileSync(process.env.WEBSOCKETS_SSL_LOCAL_PK), // /path/to/my/selfgenerated/ssl/private.key
        cert: fs.readFileSync(process.env.WEBSOCKETS_SSL_LOCAL_CERT), // /path/to/my/selfgenerated/ssl/certificat.crt
    },
},
 ```
If I don't do that using 'bootstrap.js' will get "The page at 'https://mysite.test/oncalls/create' was loaded over HTTPS, but requested an insecure script 'http://mysite.test:5173/@vite/client'. This request has been blocked; the content must be served over HTTPS." message.
I don't know if that will impact Filament instance of Laravel Echo but it do not load Echo correctly.

Here is my working bootstrap Echo configuration if it may help : 
```
window.Echo = new Echo({
    broadcaster: 'pusher',
    key: import.meta.env.VITE_PUSHER_APP_KEY,
    wsHost: window.location.hostname,
    wsPort: 6001,
    wssPort: 6001,
    forceTLS: true,
    enabledTransports: ['ws', 'wss'],
    cluster:import.meta.env.VITE_PUSHER_APP_CLUSTER,
});
```

**Reply by hossam-hani** [reply]
if anyone uses it in a custom page it will return Uncaught ReferenceError: Echo is not defined

until adding window.onload

    <script>
        window.onload=function(){
            Echo.channel('channel')
                .listen('event', (e) => {
                    console.log(e);
                });
        }
    </script>

**Reply by raonikhil** [reply]
yes, I am getting the same error, cant even get it to work with window.onload :|
